src/mame/misc/nabupc.cpp

```diff
@@ -0,0 +1,586 @@
+	NABUPC_OPTION_BUS_SLOT(config, "option3", m_bus, 2, bus::nabupc::option_bus_devices, nullptr);
+	NABUPC_OPTION_BUS_SLOT(config, "option4", m_bus, 3, bus::nabupc::option_bus_devices, nullptr);
+}
+
+
+//**************************************************************************
+//  MACHINE INITIALIZATION
+//**************************************************************************
+
+// Machine Start
+void nabupc_state::machine_start()
+{
+	m_leds.resolve();
+
+	m_hccauart->write_np(1);
+	m_hccauart->write_nb2(1);
+	m_hccauart->write_nb1(1);
+	m_hccauart->write_eps(0);
+	m_hccauart->write_tsb(0);
+	m_hccauart->write_cs(1);
+	m_hccauart->write_swe(0);
+
+	save_item(NAME(m_irq_in_prio));
+	save_item(NAME(m_int_lines));
+	save_item(NAME(m_porta));
+	save_item(NAME(m_portb));
+	save_item(NAME(m_control));
+	save_item(NAME(m_bios_size));
+}
+
+// Machine Reset
+void nabupc_state::machine_reset()
+{
+	m_irq_in_prio = PRIO_I0_I7;
+	m_int_lines = 0;
+	m_porta = 0;
+	m_portb &= 0x70;
+	m_control = 0;
+	m_bios_size = m_bios_sel->read() == 1 ? 0x2000 : 0x1000;
+	m_leds[0] = 1; // Power LED
+}
+
+
+//**************************************************************************
+//  Control/Status Ports
+//**************************************************************************
+
+// Control Register
+void nabupc_state::control_w(uint8_t data)
+{
+	m_control = data;
+	m_centronics->write_strobe(BIT(m_control, 2));
+	for (int i = 3 ; i < 6 ; ++i) {
+		m_leds[6 - i] = BIT(m_control, i);
+	}
+}
+
+// Status Port
+uint8_t nabupc_state::psg_portb_r()
+{
+	return m_portb;
+}
+
+// Interrupt Enable Port
+void nabupc_state::psg_porta_w(uint8_t data)
+{
+	if (data != m_porta) {
+		m_porta = data;
+		update_irq();
+	}
+}
+
+// Printer Busy State
+void nabupc_state::centronics_busy_handler(uint8_t state)
+{
+	BIT_SET(m_portb, 4, state);
+}
+
+// HCCA Framing Error
+WRITE_LINE_MEMBER(nabupc_state::hcca_fe_w)
+{
+	BIT_SET(m_portb, 5, state);
+}
+
+// HCCA Overrun Error
+WRITE_LINE_MEMBER(nabupc_state::hcca_oe_w)
+{
+	BIT_SET(m_portb, 6, state);
+}
+
+
+//**************************************************************************
+//  Interrupt Handling
+//**************************************************************************
+
+// IRQ For HCCA Receive
+WRITE_LINE_MEMBER(nabupc_state::hcca_dr_w)
+{
+	BIT_SET(m_int_lines, 7, state);
+	update_irq();
+}
+
+// IRQ For HCCA Transmit
+WRITE_LINE_MEMBER(nabupc_state::hcca_tbre_w)
+{
+	BIT_SET(m_int_lines, 6, state);
+	update_irq();
+}
+
+// IRQ For Keyboard
+WRITE_LINE_MEMBER(nabupc_state::rxrdy_w)
+{
+	BIT_SET(m_int_lines, 5, state);
+	update_irq();
+}
+
+// IRQ For TMS9918a
+WRITE_LINE_MEMBER(nabupc_state::vdp_int_w)
+{
+	BIT_SET(m_int_lines, 4, state);
+	update_irq();
+}
+
+// IRQ For Option Slot 1
+WRITE_LINE_MEMBER(nabupc_state::j9_int_w)
+{
+	BIT_SET(m_int_lines, 3, !state);
+	update_irq();
+}
+
+// IRQ For Option Slot 2
+WRITE_LINE_MEMBER(nabupc_state::j10_int_w)
+{
+	BIT_SET(m_int_lines, 2, !state);
+	update_irq();
+}
```

Comment:
These all seem to be identical besides the bit number.  Why not template it?  Something like:
```c++
template <unsigned N>
WRITE_LINE_MEMBER(nabupc_state::int_w)
{
	BIT_SET(m_int_lines, N, state);
	update_irq();
}
```

You can do the inversion when hooking it up in machine configuration, like:
```c++
	m_bus->out_int_callback<1>().set(FUNC(nabupc_state::int_w<2>)).invert();
```